mongodbatlas/fw_resource_mongodbatlas_project.go

```diff
@@ -0,0 +1,857 @@
+type tfProjectRSModel struct {
+	Limits                                      types.Set    `tfsdk:"limits"`
+	Teams                                       types.Set    `tfsdk:"teams"`
+	RegionUsageRestrictions                     types.String `tfsdk:"region_usage_restrictions"`
+	Name                                        types.String `tfsdk:"name"`
+	OrgID                                       types.String `tfsdk:"org_id"`
+	Created                                     types.String `tfsdk:"created"`
+	ProjectOwnerID                              types.String `tfsdk:"project_owner_id"`
+	ID                                          types.String `tfsdk:"id"`
+	ClusterCount                                types.Int64  `tfsdk:"cluster_count"`
+	IsDataExplorerEnabled                       types.Bool   `tfsdk:"is_data_explorer_enabled"`
+	IsPerformanceAdvisorEnabled                 types.Bool   `tfsdk:"is_performance_advisor_enabled"`
+	IsRealtimePerformancePanelEnabled           types.Bool   `tfsdk:"is_realtime_performance_panel_enabled"`
+	IsSchemaAdvisorEnabled                      types.Bool   `tfsdk:"is_schema_advisor_enabled"`
+	IsExtendedStorageSizesEnabled               types.Bool   `tfsdk:"is_extended_storage_sizes_enabled"`
+	IsCollectDatabaseSpecificsStatisticsEnabled types.Bool   `tfsdk:"is_collect_database_specifics_statistics_enabled"`
+	WithDefaultAlertsSettings                   types.Bool   `tfsdk:"with_default_alerts_settings"`
+}
+
+type tfTeamModel struct {
+	TeamID    types.String `tfsdk:"team_id"`
+	RoleNames types.Set    `tfsdk:"role_names"`
+}
+
+type tfLimitModel struct {
+	Name         types.String `tfsdk:"name"`
+	Value        types.Int64  `tfsdk:"value"`
+	CurrentUsage types.Int64  `tfsdk:"current_usage"`
+	DefaultLimit types.Int64  `tfsdk:"default_limit"`
+	MaximumLimit types.Int64  `tfsdk:"maximum_limit"`
+}
+
+var tfTeamObjectType = types.ObjectType{AttrTypes: map[string]attr.Type{
+	"team_id":    types.StringType,
+	"role_names": types.SetType{ElemType: types.StringType},
+}}
+var tfLimitObjectType = types.ObjectType{AttrTypes: map[string]attr.Type{
+	"name":          types.StringType,
+	"value":         types.Int64Type,
+	"current_usage": types.Int64Type,
+	"default_limit": types.Int64Type,
+	"maximum_limit": types.Int64Type,
+}}
+
+// Resources that need to be cleaned up before a project can be deleted
+type AtlastProjectDependents struct {
+	AdvancedClusters *matlas.AdvancedClustersResponse
+}
+
+func (r *ProjectRS) Metadata(ctx context.Context, req resource.MetadataRequest, resp *resource.MetadataResponse) {
+	resp.TypeName = req.ProviderTypeName + "_project"
+}
+
+func (r *ProjectRS) Configure(ctx context.Context, req resource.ConfigureRequest, resp *resource.ConfigureResponse) {
+	if req.ProviderData == nil {
+		return
+	}
+	client, ok := req.ProviderData.(*MongoDBClient)
+
+	if !ok {
+		resp.Diagnostics.AddError(errorConfigureSummary, fmt.Sprintf(errorConfigure, req.ProviderData))
+		return
+	}
+	r.client = client
+}
```

Comment:
this seems used also by other ds and resources. Can we extract the implementation in another function and reuse it inside the Configure?

Reply:
Can change it like this, didn't seem to add a lot because of the error handling that's why didn't do it initially. But let me know.

```suggestion
func (r *ProjectRS) Configure(ctx context.Context, req resource.ConfigureRequest, resp *resource.ConfigureResponse) {
	client, err := configure(req.ProviderData)
	if err != nil {
		resp.Diagnostics.AddError(errorConfigureSummary, err.Error())
		return
	}
	r.client = client
}

func configure(providerData any) (*MongoDBClient, error) {
	if providerData == nil {
		return nil, fmt.Errorf("ProviderData is null")
	}
	client, ok := providerData.(*MongoDBClient)
	if !ok {
		return nil, fmt.Errorf(errorConfigure, providerData)
	}
	return client, nil
}
```